File: 6. ZigZag Conversion/solution.cpp
/*
  Runtime: 12 ms, faster than 96.46% of C++ online submissions for ZigZag Conversion.
  Memory Usage: 12.8 MB, less than 65.01% of C++ online submissions for ZigZag 
*/

class Solution {
public:
    string convert(string s, size_t numRows) {
        if(numRows == 1)
        return s;

        vector<string> v(min(s.size(), numRows));
        size_t curRow = 0;
        bool isDown = true;
        for(const auto& c : s){
            v[curRow] += c;
            if(curRow + 1 == numRows){
                isDown = false;
            } else if(curRow == 0){
                isDown = true;
            }
            isDown ? ++curRow : --curRow;
        }

        string res;
        for(const auto& str : v){
            res += str;
        }
        return res;
    }
};

File: 6. ZigZag Conversion/solution.py
# Runtime: 44 ms, faster than 97.09% of Python online submissions for ZigZag Conversion.
# Memory Usage: 12 MB, less than 19.54% of Python online submissions for ZigZag 

class Solution(object):
    def convert(self, s, numRows):
        if numRows == 1: return s
        
        v = [''] * numRows
        isDown = 1
        curRow = 0
        for c in s:
            v[curRow] += c
            if curRow == 0:
                isDown = 1
            elif curRow + 1 == numRows:
                isDown = -1
            curRow += isDown    
        return ''.join(v)
        
